crypto/hpke.c

```diff
@@ -0,0 +1,2730 @@
+    size_t        Npk; /**< length of public key */
+    size_t        Npriv; /**< length of raw private key */
+} hpke_kem_info_t;
+
+/*!
+ * @brief table of KEMs
+ *
+ * Ok we're wasting space here, but not much and it's ok
+ */
+static hpke_kem_info_t hpke_kem_tab[] = {
+    { 0, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 1, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 2, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 3, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 4, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 5, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 6, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 7, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 8, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { 9, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {10, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {11, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {12, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {13, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {14, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {15, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { HPKE_KEM_ID_P256,
+      "EC", "P-256",
+      NID_X9_62_prime256v1, EVP_sha256,
+      32, 65, 65, 32 }, /* maybe "prime256v1" instead of P-256? */
+    { HPKE_KEM_ID_P384,
+      "EC", "P-384",
+      NID_secp384r1, EVP_sha384,
+      48, 97, 97, 48 },
+    { HPKE_KEM_ID_P521,
+      "EC", "P-521",
+      NID_secp521r1, EVP_sha512,
+      64, 133, 133, 66 },
+    {19, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {20, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {21, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {22, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {23, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {24, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {25, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {26, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {27, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {28, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {29, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {30, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    {31, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
+    { HPKE_KEM_ID_25519,
+      "X25519", NULL,
+      EVP_PKEY_X25519, EVP_sha256,
+      32, 32, 32, 32 },
+    { HPKE_KEM_ID_448,
+      "X448", NULL,
+      EVP_PKEY_X448, EVP_sha512,
+      64, 56, 56, 56 },
+    {34, NULL, NULL, 0, NULL, 0, 0, 0 }, /* keep indexing correct */
```

Comment:
If you use the ID as index, why bother storing the ID in the structure itself?

Reply:
been on my list to improve that for a while so that even much larger IANA codepoints won't cause silliness, will note here when that's done

Reply:
That one's fixed in https://github.com/openssl/openssl/pull/17172/commits/f1fc8b8e04540681cfbb647b8596afe6beb3d8ea - history ('case it helps) is that the IANA codepoints were initially sequential starting at 1 but then got changed (e.g. so x25519 was 0x20) but I hadn't gotten around to "packing" those tables. Have done that now so the structure makes more sense. (As an aside, I found including the codepoint in the struct helped a teeny bit when debugging even when the codepoint was the same as the index.) But the main reason for this (probably overdue) change is in case IANA allocate something like "0xf00d" as a code-point for some reason.